File: problems/0138-copy-list-with-random-pointer/go/solution.go
package solution

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

// Copy List with Random Pointer — https://leetcode.com/problems/copy-list-with-random-pointer/
func copyRandomList(head *Node) *Node {
	if head == nil {
		return nil
	}
	clones := map[*Node]*Node{}
	for cur := head; cur != nil; cur = cur.Next {
		clones[cur] = &Node{Val: cur.Val}
	}
	for cur := head; cur != nil; cur = cur.Next {
		clones[cur].Next = clones[cur.Next]     // clones[nil] is nil
		clones[cur].Random = clones[cur.Random] // clones[nil] is nil
	}
	return clones[head]
}

func copyRandomListArr(data [][]*int) [][]*int {
	n := len(data)
	if n == 0 {
		return [][]*int{}
	}
	nodes := make([]*Node, n)
	for i := 0; i < n; i++ {
		nodes[i] = &Node{Val: *data[i][0]}
	}
	for i := 0; i < n; i++ {
		if i+1 < n {
			nodes[i].Next = nodes[i+1]
		}
		if data[i][1] != nil {
			nodes[i].Random = nodes[*data[i][1]]
		}
	}
	copied := copyRandomList(nodes[0])
	index := map[*Node]int{}
	i := 0
	for cur := copied; cur != nil; cur = cur.Next {
		index[cur] = i
		i++
	}
	out := [][]*int{}
	for cur := copied; cur != nil; cur = cur.Next {
		v := cur.Val
		var rnd *int
		if cur.Random != nil {
			r := index[cur.Random]
			rnd = &r
		}
		out = append(out, []*int{&v, rnd})
	}
	return out
}

File: problems/0138-copy-list-with-random-pointer/python/solution.py
class Node:
    def __init__(self, val, next=None, random=None):
        self.val = val
        self.next = next
        self.random = random


def copyRandomList(head):
    """Copy List with Random Pointer — https://leetcode.com/problems/copy-list-with-random-pointer/"""
    if not head:
        return None
    clones = {}
    cur = head
    while cur:
        clones[cur] = Node(cur.val)
        cur = cur.next
    cur = head
    while cur:
        clones[cur].next = clones[cur.next] if cur.next else None
        clones[cur].random = clones[cur.random] if cur.random else None
        cur = cur.next
    return clones[head]


def copyRandomListArr(data):
    if not data:
        return []
    nodes = [Node(item[0]) for item in data]
    for i, item in enumerate(data):
        if i + 1 < len(data):
            nodes[i].next = nodes[i + 1]
        if item[1] is not None:
            nodes[i].random = nodes[item[1]]
    copied = copyRandomList(nodes[0])
    index = {}
    cur = copied
    i = 0
    while cur:
        index[id(cur)] = i
        cur = cur.next
        i += 1
    out = []
    cur = copied
    while cur:
        rnd = index[id(cur.random)] if cur.random else None
        out.append([cur.val, rnd])
        cur = cur.next
    return out
